### PR TITLE
Proposal to add native transactions

### PR DESCRIPTION
**Pull request by vkagamlyk**

### CODE REVIEW

File: docs/src/dev/future/proposal-transaction-4.asciidoc [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+* Need to reuse existing code as much as possible.
+* Users should be able to work both with and without transactions.
+* Transaction implementation should not have additional external dependencies.
+* Existing drivers should work without changes.
+
+=== Specifications
+==== Design Overview
+The main changes relate to TinkerGraph, we need to make a similar solution but with transaction support. I think TinkerGraph should be unchanged because for many users performance will be more important.
+I think by default I can make 2 configurations on Gremlin Server and users themselves will choose what they need, TinkerGraph or TinkerTransactionGraph.
```
Comment (Cole-Greer) [inline docs/src/dev/future/proposal-transaction-4.asciidoc:39]: Will users be able to run an embedded TinkerTransactionGraph?

Comment (vkagamlyk) [reply docs/src/dev/future/proposal-transaction-4.asciidoc:39]: yes, both `TinkerTransactionGraph` and `TinkerGraph` should give similar capabilities

File: docs/src/dev/future/proposal-transaction-4.asciidoc [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+
+==== CRUD operations without transaction
+1. Wrap into transaction
+2. Commit
+
+
+==== CRUD operations with transaction
+===== Create
+Add new ElementContainer to `vertices` in TinkerTransactionGraph.
```
[the remark this answers is not in the document]
Comment (vkagamlyk) [reply docs/src/dev/future/proposal-transaction-4.asciidoc:74]: yes, I mean the same solution for Vertices and Edges

File: docs/src/dev/future/proposal-transaction-4.asciidoc [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+////
+image::apache-tinkerpop-logo.png[width=500,link="https://tinkerpop.apache.org"]
+
+*x.y.z - Proposal 4*
+
+== Tinkerpop native transaction support
```
Comment (spmallette) [inline docs/src/dev/future/proposal-transaction-4.asciidoc:21]: I think the perspective this proposal is coming from is more about "TinkerGraph Transaction Support" than anything general to TinkerPop itself.

File: docs/src/dev/future/proposal-transaction-4.asciidoc [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+=== Introduction
```
Comment (spmallette) [inline docs/src/dev/future/proposal-transaction-4.asciidoc:23]: I think there's more to be said here and in "Motivation" about this. The neo4j point has validity but there's a variety of other drivers when I think about what this feature is. Test framework simplification, a more robust Gremlin Server setup out-of-the-box, a better learning experience for new users, a less complicated transaction story for new users, a better test graph that is more readily swappable with other graphs, and probably more. Please expand these sections further. 

Also, please get "TinkerPop" spelling corrections in place. 🙂

File: docs/src/dev/future/proposal-transaction-4.asciidoc [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+class ElementContainer {
+    Vertex current;
+    ThreadLocal<Vertex> transactionUpdatedValue;
+    ThreadLocal<Boolean> isDeleted;
+}
+----
+
+===== TinkerTransactionGraph
+Vertices and Edges will store with version number and dirty transaction values
+`Map<Object, ElementContainer> vertices` where `Object` is Vertex Id.
```
[the remark this answers is not in the document]
Comment (vkagamlyk) [reply docs/src/dev/future/proposal-transaction-4.asciidoc:61]: yes, same for Vertixes and Edges

File: docs/src/dev/future/proposal-transaction-4.asciidoc (hunk 3 above)
Comment (vkagamlyk) [reply docs/src/dev/future/proposal-transaction-4.asciidoc]: updated in https://github.com/apache/tinkerpop/pull/2054/commits/4f3f7e916cf4b44db550c67663bd1f4d55bac598

File: docs/src/dev/future/proposal-transaction-4.asciidoc (hunk 4 above)
Comment (vkagamlyk) [reply docs/src/dev/future/proposal-transaction-4.asciidoc:23]: updated in https://github.com/apache/tinkerpop/pull/2054/commits/4f3f7e916cf4b44db550c67663bd1f4d55bac598

**Pull request merged**